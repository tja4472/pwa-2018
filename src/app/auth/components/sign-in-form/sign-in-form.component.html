<form [formGroup]="loginForm"
      (ngSubmit)="onSubmit()">
  <ion-text>
    Email Address: a.a@a.com
    <br> Password: password
  </ion-text>
  <ion-list lines="none">
    <ion-item>
      <ion-label position="stacked"
                 color="primary">Email address</ion-label>
      <ion-input formControlName="username"
                 placeholder="Username"
                 type="email"
                 autofocus
                 required></ion-input>
      <div *ngIf="username.invalid  && (username.dirty || username.touched)">
        <div *ngIf="username.hasError('required')">
          Name is required.
        </div>
      </div>
    </ion-item>

    <ion-item>
      <ion-label position="stacked"
                 color="primary">Password</ion-label>
      <ion-input type="password"
                 formControlName="password"
                 placeholder="Password"
                 required></ion-input>
      <div *ngIf="password.invalid  && (password.dirty || password.touched)">
        <div *ngIf="password.hasError('required')">
          Password is required.
        </div>
      </div>
    </ion-item>

    <ion-text data-test="error-message"
              *ngIf="errorMessage"
              color="danger">
      {{errorMessage}}
      <!--
        <p [hidden]="username.valid || submitted == false" padding-left>
          Username is required
        </p>
      -->
    </ion-text>
  </ion-list>

  <ion-grid>
    <ion-row>
      <ion-col>
        <ion-button type="submit"
                    [disabled]="!loginForm.valid"
                    expand="block">Sign In</ion-button>
      </ion-col>
      <ion-col>
        <ion-button (click)="signUp()"
                    color="light"
                    expand="block">Sign up</ion-button>
      </ion-col>
    </ion-row>
  </ion-grid>
</form>

<!--
<p>
  Form Value: {{ loginForm.value | json }}
</p>

<p>
  Form Status: {{ loginForm.status }}
</p>
-->