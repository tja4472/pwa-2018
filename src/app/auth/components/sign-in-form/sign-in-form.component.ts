import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';

import { FormControl, FormGroup, Validators } from '@angular/forms';

import { Credentials } from '@app/auth/models/credentials.model';

@Component({
  selector: 'tja-sign-in-form',
  templateUrl: './sign-in-form.component.html',
  styleUrls: ['./sign-in-form.component.css'],
})
export class SignInFormComponent implements OnInit {
  @Input()
  errorMessage: string | null;

  @Input()
  set pending(isPending: boolean) {
    if (isPending) {
      this.loginForm.disable();
    } else {
      this.loginForm.enable();
    }
  }

  @Output()
  readonly SignUpClicked = new EventEmitter();

  @Output()
  readonly submitted = new EventEmitter<Credentials>();

  get password() {
    return this.loginForm.get('password');
  }
  get username() {
    return this.loginForm.get('username');
  }

  public loginForm = new FormGroup({
    username: new FormControl('', [Validators.required]),
    password: new FormControl('', [Validators.required]),
  });

  constructor() {}

  ngOnInit() {}

  signUp() {
    this.SignUpClicked.emit();
  }

  onSubmit() {
    //
    const value: Credentials = this.loginForm.value;

    if (this.loginForm.valid) {
      this.submitted.emit(value);
    }
  }
}
