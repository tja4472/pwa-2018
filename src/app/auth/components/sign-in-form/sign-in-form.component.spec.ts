import {
  Location,
  LocationStrategy,
  PathLocationStrategy,
} from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';

import { IonicModule } from '@ionic/angular';

import { SignInFormComponent } from '@app/auth/components/sign-in-form/sign-in-form.component';

describe('SignInFormComponent - class tests', () => {
  it('raises the submitted event when valid form submitted', () => {
    const component = new SignInFormComponent();
    const credentials = { username: 'a.a@a.com', password: '123456789' };

    component.loginForm.setValue(credentials);
    expect(component.loginForm.valid).toBeTruthy();

    component.submitted.subscribe((x) => expect(x).toEqual(credentials));
    component.onSubmit();
  });

  it('should disable the form if pending', () => {
    const component = new SignInFormComponent();
    component.pending = true;
    expect(component.loginForm.disabled).toBeTruthy();
  });

  it('should enable the form if not pending', () => {
    const component = new SignInFormComponent();
    component.pending = false;
    expect(component.loginForm.disabled).toBeFalsy();
  });
});

describe('SignInFormComponent - DOM tests', () => {
  let component: SignInFormComponent;
  let fixture: ComponentFixture<SignInFormComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [IonicModule.forRoot(), ReactiveFormsModule],
      declarations: [SignInFormComponent],
      providers: [
        Location,
        { provide: LocationStrategy, useClass: PathLocationStrategy },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SignInFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeDefined();
  });

  it('should disable the form if pending', () => {
    component.pending = true;
    fixture.detectChanges();
    expect(component.loginForm.disabled).toBeTruthy();
  });

  it('should display an error message if provided', () => {
    const componentEl: HTMLElement = fixture.nativeElement;
    component.errorMessage = 'Invalid credentials';
    fixture.detectChanges();
    const errorMessage = componentEl.querySelector(
      '[data-test="error-message"]'
    );
    expect(errorMessage.textContent).toContain(component.errorMessage);
  });

  it('should emit an event if the form is valid when submitted', () => {
    const credentials = {
      username: 'user',
      password: 'pass',
    };
    component.loginForm.setValue(credentials);

    spyOn(component.submitted, 'emit');
    component.onSubmit();

    expect(component.submitted.emit).toHaveBeenCalledWith(credentials);
  });

  it('dummy', () => {
    expect(1).toEqual(1);
  });
});
