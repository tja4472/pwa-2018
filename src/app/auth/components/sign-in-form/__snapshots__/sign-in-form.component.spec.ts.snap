// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In Page ngrx-platform - should compile 1`] = `
<tja-sign-in-form
  SignUpClicked={[Function EventEmitter]}
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <ion-text>
       Email Address: a.a@a.com 
      <br />
       Password: password 
    </ion-text>
    <ion-list
      lines="none"
      ng-reflect-lines="none"
    >
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Email address
        </ion-label>
        <ion-input
          autofocus=""
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="username"
          ng-reflect-autofocus=""
          ng-reflect-name="username"
          ng-reflect-placeholder="Username"
          ng-reflect-required=""
          ng-reflect-type="email"
          placeholder="Username"
          required=""
          type="email"
        />
        
      </ion-item>
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Password
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="password"
          ng-reflect-name="password"
          ng-reflect-placeholder="Password"
          ng-reflect-required=""
          ng-reflect-type="password"
          placeholder="Password"
          required=""
          type="password"
        />
        
      </ion-item>
      
    </ion-list>
    <ion-grid>
      <ion-row>
        <ion-col>
          <ion-button
            expand="block"
            ng-reflect-disabled="true"
            ng-reflect-expand="block"
            ng-reflect-type="submit"
            type="submit"
          >
            Sign In
          </ion-button>
        </ion-col>
        <ion-col>
          <ion-button
            color="light"
            expand="block"
            ng-reflect-color="light"
            ng-reflect-expand="block"
          >
            Sign up
          </ion-button>
        </ion-col>
      </ion-row>
    </ion-grid>
  </form>
</tja-sign-in-form>
`;

exports[`Sign In Page ngrx-platform - should disable the form if pending 1`] = `
<tja-sign-in-form
  SignUpClicked={[Function EventEmitter]}
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <ion-text>
       Email Address: a.a@a.com 
      <br />
       Password: password 
    </ion-text>
    <ion-list
      lines="none"
      ng-reflect-lines="none"
    >
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Email address
        </ion-label>
        <ion-input
          autofocus=""
          class="ng-untouched ng-pristine"
          formcontrolname="username"
          ng-reflect-autofocus=""
          ng-reflect-name="username"
          ng-reflect-placeholder="Username"
          ng-reflect-required=""
          ng-reflect-type="email"
          placeholder="Username"
          required=""
          type="email"
        />
        
      </ion-item>
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Password
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine"
          formcontrolname="password"
          ng-reflect-name="password"
          ng-reflect-placeholder="Password"
          ng-reflect-required=""
          ng-reflect-type="password"
          placeholder="Password"
          required=""
          type="password"
        />
        
      </ion-item>
      
    </ion-list>
    <ion-grid>
      <ion-row>
        <ion-col>
          <ion-button
            expand="block"
            ng-reflect-disabled="true"
            ng-reflect-expand="block"
            ng-reflect-type="submit"
            type="submit"
          >
            Sign In
          </ion-button>
        </ion-col>
        <ion-col>
          <ion-button
            color="light"
            expand="block"
            ng-reflect-color="light"
            ng-reflect-expand="block"
          >
            Sign up
          </ion-button>
        </ion-col>
      </ion-row>
    </ion-grid>
  </form>
</tja-sign-in-form>
`;

exports[`Sign In Page ngrx-platform - should display an error message if provided 1`] = `
<tja-sign-in-form
  SignUpClicked={[Function EventEmitter]}
  errorMessage={[Function String]}
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <ion-text>
       Email Address: a.a@a.com 
      <br />
       Password: password 
    </ion-text>
    <ion-list
      lines="none"
      ng-reflect-lines="none"
    >
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Email address
        </ion-label>
        <ion-input
          autofocus=""
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="username"
          ng-reflect-autofocus=""
          ng-reflect-name="username"
          ng-reflect-placeholder="Username"
          ng-reflect-required=""
          ng-reflect-type="email"
          placeholder="Username"
          required=""
          type="email"
        />
        
      </ion-item>
      <ion-item>
        <ion-label
          color="primary"
          ng-reflect-color="primary"
          ng-reflect-position="stacked"
          position="stacked"
        >
          Password
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid"
          formcontrolname="password"
          ng-reflect-name="password"
          ng-reflect-placeholder="Password"
          ng-reflect-required=""
          ng-reflect-type="password"
          placeholder="Password"
          required=""
          type="password"
        />
        
      </ion-item>
      
      <ion-text
        color="danger"
        ng-reflect-color="danger"
      >
         Invalid credentials 
      </ion-text>
    </ion-list>
    <ion-grid>
      <ion-row>
        <ion-col>
          <ion-button
            expand="block"
            ng-reflect-disabled="true"
            ng-reflect-expand="block"
            ng-reflect-type="submit"
            type="submit"
          >
            Sign In
          </ion-button>
        </ion-col>
        <ion-col>
          <ion-button
            color="light"
            expand="block"
            ng-reflect-color="light"
            ng-reflect-expand="block"
          >
            Sign up
          </ion-button>
        </ion-col>
      </ion-row>
    </ion-grid>
  </form>
</tja-sign-in-form>
`;
